base/array.jl

```diff
@@ -196,10 +228,34 @@ end
 
 convert{T,n}(::Type{Array{T}}, x::Array{T,n}) = x
 convert{T,n}(::Type{Array{T,n}}, x::Array{T,n}) = x
+# Defined to prevent ambiguities
+convert{T}(::Type{Vector{T}}, x::Vector{T}) = x
+convert{T}(::Type{Matrix{T}}, x::Matrix{T}) = x
 
 convert{T,n,S}(::Type{Array{T}}, x::AbstractArray{S, n}) = convert(Array{T, n}, x)
 convert{T,n,S}(::Type{Array{T,n}}, x::AbstractArray{S,n}) = copy!(Array{T}(size(x)), x)
 
+convert{S}(::Type{Vector}, x::AbstractArray{S,1}) = convert(Vector{S}, x)
+convert{S}(::Type{Matrix}, x::AbstractArray{S,2}) = convert(Matrix{S}, x)
+
+function _check_itr_dim(itr, n)
+    if !isa(iteratorsize(itr), HasShape)
+        throw(DimensionMismatch("cannot convert iterator of type $(typeof(itr)) with unknown size to an Array{T, $n}"))
+    elseif ndims(itr) != n
+        throw(DimensionMismatch("cannot convert iterator of type $(typeof(itr)) with size $(size(itr)) to an Array{T, $n}"))
+    end
+end
+
+convert(::Type{Array}, itr::Any) = collect(itr)
+convert{T}(::Type{Array{T}}, itr::Any) = collect(T, itr)
+convert{T,n}(::Type{Array{T,n}}, itr::Any) = (_check_itr_dim(itr, n); collect(T, itr))
+
```

Comment:
In this file the dimention is usually `N` (when a type parameter) except for those few `convert` methods, you could convert `n` to `N` here. Any reason here to dispatch on `Any` explicitly for `itr`?